crates/wiggle/src/borrow.rs

```diff
@@ -1,64 +1,94 @@
+        self.bc.borrow_mut().borrow(r)
     }
+    pub(crate) fn unborrow(&self, h: BorrowHandle) {
+        self.bc.borrow_mut().unborrow(h)
+    }
+    pub(crate) fn is_borrowed(&self, r: Region) -> bool {
+        self.bc.borrow().is_borrowed(r)
+    }
+}
 
-    pub(crate) fn borrow(&mut self, r: Region) -> Result<(), GuestError> {
-        if self.is_borrowed(r) {
-            Err(GuestError::PtrBorrowed(r))
-        } else {
-            self.borrows.push(r);
-            Ok(())
+#[derive(Debug)]
+struct InnerBorrowChecker {
+    borrows: HashMap<BorrowHandle, Region>,
+    next_handle: BorrowHandle,
+}
+
+impl InnerBorrowChecker {
+    fn new() -> Self {
+        InnerBorrowChecker {
+            borrows: HashMap::new(),
+            next_handle: BorrowHandle(0),
         }
     }
 
-    /// Borrow the region of memory pointed to by a `GuestPtr`. This is required for safety if
-    /// you are dereferencing `GuestPtr`s while holding a reference to a slice via
-    /// `GuestPtr::as_raw`.
-    pub fn borrow_pointee<'a, T>(&mut self, p: &GuestPtr<'a, T>) -> Result<(), GuestError>
-    where
-        T: GuestType<'a>,
-    {
-        self.borrow(Region {
-            start: p.offset(),
-            len: T::guest_size(),
-        })
+    fn has_outstanding_borrows(&self) -> bool {
+        !self.borrows.is_empty()
     }
 
-    /// Borrow the slice of memory pointed to by a `GuestPtr<[T]>`. This is required for safety if
-    /// you are dereferencing the `GuestPtr`s while holding a reference to another slice via
-    /// `GuestPtr::as_raw`. Not required if using `GuestPtr::as_raw` on this pointer.
-    pub fn borrow_slice<'a, T>(&mut self, p: &GuestPtr<'a, [T]>) -> Result<(), GuestError>
-    where
-        T: GuestType<'a>,
-    {
-        let (start, elems) = p.offset();
-        let len = T::guest_size()
-            .checked_mul(elems)
-            .ok_or_else(|| GuestError::PtrOverflow)?;
-        self.borrow(Region { start, len })
+    fn is_borrowed(&self, r: Region) -> bool {
+        !self.borrows.values().all(|b| !b.overlaps(r))
     }
 
-    /// Borrow the slice of memory pointed to by a `GuestPtr<str>`. This is required for safety if
-    /// you are dereferencing the `GuestPtr`s while holding a reference to another slice via
-    /// `GuestPtr::as_raw`. Not required if using `GuestPtr::as_raw` on this pointer.
-    pub fn borrow_str(&mut self, p: &GuestPtr<str>) -> Result<(), GuestError> {
-        let (start, len) = p.offset();
-        self.borrow(Region { start, len })
+    fn new_handle(&mut self) -> Result<BorrowHandle, GuestError> {
+        // Reset handles to 0 if all handles have been returned.
+        if self.borrows.is_empty() {
+            self.next_handle = BorrowHandle(0);
+        }
+        let h = self.next_handle;
+        self.next_handle = BorrowHandle(
+            h.0.checked_add(1)
+                .ok_or_else(|| GuestError::BorrowCheckerOOM)?,
+        );
+        Ok(h)
```

Comment:
I might have missed it, but do we actually handle the case here where, if we handed out all available handles, but then the user unborrows some region, we actually still trip with `BorrowCheckerOOM` on the next call to `self.new_handle()` since the unborrowed handle is never returned back to the pool? Or do we actually take care of it and I just missed it?

Reply:
Oh and btw, it might also be good to add this scenario as a test case.

Reply:
Yes, this datastructure isnt smart enough to recycle handles, because I didn't want to incur the bookkeeping cost for what should only happen in a pathological scenario of user code - if you were to `{ borrow(r1); loop { h2 = borrow(r2); unborrow(h2) } }` you'd eventually run out of handles. I can add a test that shows this bad behavior is present, with a note that we might want to try to fix it.

If we recycled all unborrowed handles, it would be impossible to run out because the bound of `usize` handles is at least as big as the 32 bit memory space described by `Region`.

Reply:
Yep, that's a good point on the number of handles. I noticed this since we've faced the same problem with the `FdPool` in `wasi-common`, there handles being actual WASI file descriptors. Anyhow, a test case demonstrating the behaviour plus a comment sounds great to me!

Reply:
I wrote a test for this (requires 2^64 iterations) and it ran for more than 60 seconds before I killed it, so I don't want to include it in the test suite. I'll leave a description of the behavior as a comment.

Reply:
Oh sorry, I thought you’d spoof the test by doing a borrow, followed by an unborrow, and then manually fast forward next available handle to the limit and assert this fails whereas it shouldn’t since there was at least one unborrow registered. Having said that I reckon a comment will do nicely for now.